### ISSUE TITLE
Incorporating Trade Mechanics

### ISSUE THREAD

**Issue by JeremyWildsmith** [JeremyWildsmith/bazaarbot-java#3]
I want to find a way to incorporate trade mechanics.

Right now, when two agents make a trade, the goods are immediately transferred between the two of them.

I wonder what changes would be necessary to allow for a little more complex trade mechanics. Ie, introduce the cost of delivering the goods. Some ideas:

1. We could have the two parties form a trade contract (agreement to exchange goods before a particular dead line)
2. A bid is made to a transporation service market to fulfill the delivery
3. If the delivery bid is not accepted or cannot be made before the agreed deadline, it is cancelled and has to be reformed.
4. Different agents in the trade transportation market would need to hold price beliefs of the service (maybe with the cost of travel as a base cost?)
5. Right now, agents hold price beliefs of their goods; in the above scenario it would need to be specific not just to goods, but to agents as well. Ie price belief of trading iron _with_ agent x

A couple problems I see with this:

- It probably isn't the best approach to cancel contracts if they cannot be deliivered. Instead, it would be better if the trade contract was formed with the transportation contract. Not sure how to do this.
- How to we determine how long a trade delivery period is acceptable?
- We need to factor in "goods that are expected to be retrieved" into the agent's logic. If they order wood, and then find out in the next round "Oh I'm low on wood" and orders wood again, that wouldn't make sense.

I am open to input.

Thanks!

Jeremy Wildsmith

**Comment by JeremyWildsmith** [comment]
I think we should wait until after refactoring until implementation, but I think the implementation will be pretty simple. I'm just not sure about the algorithm to follow for introducing these mechanics?

**Comment by JeremyWildsmith** [comment]
FYI this is where the classes under contract came from; and where the _expected field in the Inventory class came from. They currently aren't doing anything right now though.

**Comment by DigitalSmile** [comment]
Yep, I would agree, let's do some refactoring first. Will do some changes today.

**Comment by JeremyWildsmith** [comment]
Great, sounds good.

Btw feel free to open an issue if you want to discus any refactors I made or explore better alternatives.

Thanks!

**Comment by JeremyWildsmith** [comment]
I have refactored the code to allow injection of custom offer execution and resolution mechanics. This should allow for contract integration a little later.
